README.md
<h1 align="center">
   AnimateSprite
</h1>

![npm (scoped)](https://img.shields.io/npm/v/@its2easy/animate-sprite)
![npm bundle size (scoped)](https://img.shields.io/bundlephobia/min/@its2easy/animate-sprite)

Demo - [codepen](https://codepen.io/its2easy/pen/VwvVRed)

**AnimateSprite** is a library (12kb) without built-in UI, which allows to animate a 
sequence of frames that looks like 3d rotation. It supports playing full animation
sequence, play a specified number of frames, rotate an image with the mouse and touch; and 
exposes an API to programmatically control the animation.

To use it, you should export animation from your 3d software as a series of 
frames and combine them into one image file (sprite). If you want to use a sequence of 
images instead of a single sprite, you could use [animate-images](https://github.com/its2easy/animate-images).

* [Installation](#installation)
* [Usage](#usage)
* [Options](#options)
* [Methods](#methods)
* [Events](#events)
* [Browser support](#browser_support)
* [License](#license)

## <a name="installation"></a>Installation
### Browser script tag
Add with CDN link:
```html
<script src="https://cdn.jsdelivr.net/npm/@its2easy/animate-sprite"></script>
```

### npm
```
npm i @its2easy/animate-sprite --save
```

## <a name="usage"></a>Usage
Create an element with a background:
```html
<div id="sprite" style="background-image: url('images/sprite.jpg')"></div>
```
Initialize with options:
```javascript
let element = document.getElementById('sprite');
let sprite = new AnimateSprite(element, {
    width: 800, /* required */
    height: 450, /* required */
    frames: 90, /* required */
    cols: 10,
    fps: 60,
    loop: true,
    draggable: true
});
sprite.play();
```
Or with inline options and then initialize it:
```html
<div id="sprite" 
    style="background-image: url('images/sprite.jpg')"
    data-sprite-width="700"
    data-sprite-heigth="400"
    data-sprite-frames="45"
    data-sprite-autoplay="true"
    data-sprite-touch-scroll-mode="allowPageScroll"
></div>
```
```javascript
let sprite = new AnimateSprite(document.getElementById('sprite'));
```
> Inline options have higher priority

### Usage with bundlers
```javascript
import AnimateSprite from '@its2easy/animate-sprite';
let sprite = new AnimateSprite(element, options);
```
It is possible to directly import untranspiled esm version:
```javascript
import AnimateSprite from '@its2easy/animate-sprite/build/untranspiled/animate-sprite.esm.js'; //or animate-sprite.esm.min.js
```
> :warning: You should probably add it to your build process if you use untranspiled version. Example for webpack:
```javascript
rules: [
    {
        test: /\.js$/,
        exclude: /node_modules(?!(\/|\\)@its2easy(\/|\\)animate-sprite(\/|\\)build)/,
        use: {
            loader: 'babel-loader',
        }
    }
]
```
or
```javascript
rules: [
    {
        // basic js rule
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
            loader: 'babel-loader',
        }
    },
    {
        // additional rule
        test: /\.js$/,
        include: /node_modules(\/|\\)@its2easy(\/|\\)animate-sprite(\/|\\)build/,
        use: {
            loader: 'babel-loader',
    }
]
```
#### All available versions:
##### umd build:

`@its2easy/animate-sprite/build/animate-sprite.umd.min.js` - default for browser script tag and legacy bundlers

##### esm builds processed whit babel:

`@its2easy/animate-sprite/build/animate-sprite.esm.min.js` - default for webpack and module environments

`@its2easy/animate-sprite/build/animate-sprite.esm.js`

##### esm builds without babel transformation:

`@its2easy/animate-sprite/build/untranspiled/animate-sprite.esm.min.js`

`@its2easy/animate-sprite/build/untranspiled/animate-sprite.esm.js`

### Responsive behavior
By default, you don't have to specify block sizes in css. The element with the
sprite will take all available width and set its height based on the width/height
ratio from the configuration options. To limit the size, add `max-width` to the
sprite. You can also explicitly set `width`, but it will no longer adapt 
to the container width (on small screens this can be changed with 
`width: auto` inside media query).

On page load sprite block will have 0px height, to prevent this you can manually
add `height` style (and maybe `backgroud-size`), it will be overwritten
after initialization. 


## <a name="options"></a>Options

| Parameter  | Type | Required | Default | Description |
| :--- | :--- | :---:| :---: | ---  |
| **width** <br> ```data-sprite-width``` | number | :heavy_check_mark:  |   | Width of one frame in a sprite  |
| **height** <br> ```data-sprite-height``` | number | :heavy_check_mark: |   | Height of one frame in a sprite  |
| **frames** <br> ```data-sprite-frames``` | number | :heavy_check_mark: |   | Total number of frames  |
| **cols** <br> ```data-sprite-cols```| number&#124;false |  | false | Number of cols if more than 1 row |
| **loop** <br> ```data-sprite-loop``` | boolean |  | false |  Loop the animation |
| **autoplay** <br> ```data-sprite-autoplay``` | boolean |  | false |  Autoplay |
| **frameTime** <br> ```data-sprite-frame-time``` | number&#124;false | | false |  ms, time between frames |
| **duration** <br> ```data-sprite-duration``` | number&#124;false |  | false |  ms, total time, alternative to frameTime |
| **fps** <br> ```data-sprite-fps``` | number&#124;false |  | 24 |  fps, alternative to frameTime |
| **reverse** <br> ```data-sprite-reverse``` | boolean |  | false |  Reverse animation direction |
| **draggable** <br> ```data-sprite-draggable``` | boolean |  | false |  Draggable by mouse or touch |
| **inversion** <br> ```data-sprite-inversion``` | boolean |  | false |  Inversion changes drag direction |
| **dragModifier** <br> ```data-sprite-drag-modifier``` | number | | 1 | Sensitivity factor for user interaction. Only positive numbers are allowed |
| **touchScrollMode** <br> ```data-sprite-touch-scroll-mode``` | string | | "pageScrollTimer" | Page scroll behavior with touch events _(only for events that fire in the plugin area)_. Available modes: `preventPageScroll` - touch scroll is always disabled. `allowPageScroll` - touch scroll is always enabled. `pageScrollTimer` - after the first interaction the scroll is not disabled; if the time between the end of the previous interaction and the start of a new one is less than _pageScrollTimerDelay_, then scroll will be disabled; if more time has passed, then scroll will be enabled again |
| **pageScrollTimerDelay** <br> ```data-sprite-page-scroll-timer-delay``` | number | | 1500 | Time in ms when touch scroll will be disabled after the last user interaction, if `touchScrollMode: "pageScrollTimer"` |
| **onAnimationEnd** | function(AnimateSprite) | | | Callback, occurs when animation has ended, receives plugin instance as a parameter |
| **onAfterFrameChange** | function(AnimateSprite) | | | Callback, occurs after the frame has switched, receives plugin instance as a parameter |

> If multiple time options (`frameTime`, `duration` or `fps`) are set at the 
same time, `frameTime` has higher priority, then `duration`, then `fps`.

## <a name="methods"></a>Methods
>  Methods can be chained ```sprite.setReverse(true).play()```
### new AnimateSprite(element, options)
Initializes and returns an instance of a sprite.

`parameters`
- element : {Element | HTMLElement} - HTML element
- options : {Object} - configuration [options](#options)

`returns` {AnimateSprite} - plugin instance

---

### play
Start animation

`returns` {AnimateSprite} - plugin instance

---

### stop
Stop animation

`returns` {AnimateSprite} - plugin instance

---

### toggle
Toggle between play and stop

`returns` {AnimateSprite} - plugin instance

---

### prev
Show the previous frame

`returns` {AnimateSprite} - plugin instance

---

### next
Show the next frame

`returns` {AnimateSprite} - plugin instance

---

### setFrame
Set frame (without animation)

`parameters`
- frameNumber : {number} - Frame number
```javascript
sprite.setFrame(35);
```
`returns` {AnimateSprite} - plugin instance

---

### playTo
Starts the animation, that plays until the specified frame number

`parameters`
- frameNumber {number} - Target frame number
- options {Object}
    - options.shortestPath {boolean} - If set to true and **loop is enabled**, function will use the shortest path to the target frame,
      even if the path crosses edge frames. Default is **false**.
```javascript
// if current frame is 30 of 100, it will play from 30 to 85, 
// if current frame is 95, it will play from 95 to 85
sprite.playTo(85);

// shortestPath
// if current frame is 2, it will play 1, 100, 99, 98
sprite.playTo(98, {
    shortestPath: true
});
// (default) if current frame is 2, it will play 3, 4, 5 ... 97, 98
sprite.playTo(98);
```
`returns` {AnimateSprite} - plugin instance

---

### playFrames
Start animation in the current direction with the specified number of frames in the queue. 
If `options.loop: false` animation will stop when it reaches the first or the last frame.

`parameters`
- numberOfFrames {number} - Number of frames to play
```javascript
instance.playFrames(200);
```
`returns` {AnimateSprite} - plugin instance

---

### setReverse
Change the direction of the animation.  Alias to ```setOption('reverse', true)```

`parameters`
- reverse : {boolean} - true to reverse, false to normal direction
```javascript
sprite.setReverse(true);
```
`returns` {AnimateSprite} - plugin instance

---


### getReverse
Returns current reverse value. Alias to getOption('reverse')

`returns` {boolean} - reverse


---

### updateSizes
Calculate new dimensions (sprite element and frame), this function should be called if element 
size was changed by a script. Called automatically after page resize

`returns` {AnimateSprite} - plugin instance

---

### getOption
Returns option value

`parameters`
- option {string} -  Option name. All options are allowed.
```javascript
let loop = instance.getOption('loop');
```
`returns` {*} - Option value

---

### setOption
Set new option value

`parameters`
- option : {string} - Option name. All options are allowed.
- value : {*} - New value
```javascript
sprite.setOption('frameTime', 40);
```
`returns` {AnimateSprite} - plugin instance

---

### getCurrentFrame
Returns current frame number

`returns` {number} - Frame number

---

### isAnimating
Returns `true` if sprite is animating

`returns` {boolean}

---


### reset
Stop the animation and return to the first frame

`returns` {AnimateSprite} - plugin instance

---

### destroy
Stop animation, remove event listeners. Method doesn't remove sprite element from the DOM

---

## <a name="events"></a>Events

#### sprite:drag-start
Fires when user starts dragging. Frame number is in `event.detail.frame`

#### sprite:drag-change
Fires on every frame change while dragging. Frame number is in `event.detail.frame`

#### sprite:drag-end
Fires when user stops dragging. Frame number is in `event.detail.frame`

#### sprite:animation-end
Fires after the animation end. If the second animation was started
while the first was active, this event will be fired only after the
second animation end.

Example:
```javascript
var element = document.getElementById('sprite');
element.addEventListener('sprite:drag-end', function (event) {
    console.log(event.detail.frame);
})
```

## <a name="browser_support"></a>Browser support
* latest versions of Chrome, android Chrome, Edge, Firefox
* Safari 13.1+,
* iOS Safari 13.4+

## <a name="license"></a>License
AnimateSprite is provided under the [MIT License](https://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog
## 2.2.1
- new ```shortestPath``` option for ```playTo()```
## 2.2.0
- new ```onAnimationEnd``` and ```onFrameChange``` callbacks
- new ```sprite:animation-end``` event
- new ```dragModifier``` option
- more validations
- fix wrong animation duration
- ```inversion``` is now only used while dragging and doesn't affect animation
- proper handling of empty animation in ```playTo``` and ```playFrames```
- fix animation stop after dragging and loop: false
- umd version filename is now ```build/animate-sprite.umd.min.js```
## 2.1.0
- add types, fix documentation
## 2.0.1
- fix package.json main key
## 2.0.0
- plugin import changed
- new initialization with constructor instead of ```init``` method  
- inline options format has changed from ```data-width``` to ```data-sprite-width```
- new options ```inversion```, ```touchScrollMode``` and ```pageScrollTimerDelay```
- new methods ```playTo```, ```playFrames```, ```getOption```, ```updateSizes``` and 
  ```getReverse``` (alias to ```getOption("reverse")```)
- all options are now allowed in ```setOption```
- all options are now allowed as inline options
- ```sprite:last-frame``` and ```sprite:first-frame``` events have been removed
- supported browsers have changed (Safari 13.1+, iOS Safari 13.4+)



src/AnimateSprite.js
import { validateParameters, validateSettings, getSettings, getSettingsKeys } from "./settings";
import { normalizeFrameNumber } from "./utils";
import DragInput from "./DragInput";
import Animation from "./Animation";

//todo core-js 3.21 adds es.error.cause(wasm polyfill?) to every file, current 3.20 is ok, check in the future
/**
 * Animate Sprite {@link https://github.com/its2easy/animate-sprite/}
 * @example
 * let sprite = new AnimateSprite( document.getElementById('sprite'), {
 *     width: 720,
 *     height: 405,
 *     frames: 20,
 *     cols: 5,
 *     fps: 60,
 * });
 */
export default class AnimateSprite {
    #settings;
    #data = {
        currentFrame: 1,
        nodeWidth: null,
        nodeHeight: null,
        element: null,
        pluginApi: {},
    }
    #boundCalculateSizes;
    // Classes
    #dragInput;
    #animation;

    /**
     * Creates plugin instance
     * @param {Element|HTMLElement} node - HTML element
     * @param {PluginOptions} options
     */
    constructor( node, options ) {
        validateParameters(node, options);
        this.#settings = getSettings(node, options);
        validateSettings(this.#settings);
        this.#data.element = node;
        this.#data.pluginApi = this;
        this.#boundCalculateSizes = this.#calculateSizes.bind(this);
        this.#initPlugin();
    }

    #initPlugin(){
        this.#animation = new Animation(
            {settings: this.#settings, data: this.#data, changeFrame: this.#changeFrame.bind(this)} );
        this.#calculateSizes();
        this.#toggleResizeHandler(true);

        if ( this.#settings.autoplay ) this.play();
        if ( this.#settings.draggable ) this.#toggleDrag(true);
    }

    #animateSprite(frame) {
        this.#data.element.style.backgroundPosition = this.#calculatePosition(frame);
        if ( this.#settings.onAfterFrameChange ) this.#settings.onAfterFrameChange(this.#data.pluginApi);
    }
    #changeFrame(frameNumber){
        if (frameNumber === this.#data.currentFrame) return;//skip same frame
        this.#animateSprite(frameNumber);
        this.#data.currentFrame = frameNumber;
    }
    #calculatePosition(frame){
        let xPadding, yPadding = 0;
        if ( !this.#settings.cols ){ // Single row sprite
            xPadding = (frame - 1) * this.#data.nodeWidth;
        } else { // Multiline sprite
            xPadding = ( ( (frame - 1) % this.#settings.cols)  ) * this.#data.nodeWidth;
            yPadding = Math.floor( (frame - 1) / this.#settings.cols ) * this.#data.nodeHeight;
        }
        return `-${xPadding}px -${yPadding}px`;
    }

    #calculateSizes(){
        const newNodeWidth = this.#data.element.offsetWidth;
        if (this.#data.nodeWidth === newNodeWidth ) return;//skip if same size

        const wasAnimating = this.#animation.isAnimating,
              widthHeightRatio = this.#settings.width / this.#settings.height;
        this.#data.nodeWidth = newNodeWidth;
        this.#data.nodeHeight = this.#data.nodeWidth / widthHeightRatio;
        this.#data.element.style.height = this.#data.nodeHeight + "px";

        let bgWidth =  ( !this.#settings.cols )
            ? this.#settings.frames * this.#data.nodeWidth
            : this.#settings.cols * this.#data.nodeWidth;
        let bgHeight = ( !this.#settings.cols )
            ? this.#data.nodeHeight
            : Math.ceil( this.#settings.frames / this.#settings.cols ) * this.#data.nodeHeight;
        this.#data.element.style.backgroundSize = `${bgWidth}px ${bgHeight}px`;
        if (wasAnimating) this.play();
            else this.#animateSprite(this.#data.currentFrame);
        if ( this.#dragInput ) this.#dragInput.updateThreshold()
    }

    #toggleDrag(enable = true){
        if (enable) {
            if ( !this.#dragInput ) this.#dragInput = new DragInput({
                data: this.#data,
                settings: this.#settings,
                changeFrame: this.#changeFrame.bind(this),
                getNextFrame: this.#animation.getNextFrame.bind(this.#animation)
            });
            this.#dragInput.enableDrag();
        } else {
            if (this.#dragInput) this.#dragInput.disableDrag();
        }
    }

    #toggleResizeHandler(add = true) {
        if ( add ) window.addEventListener("resize", this.#boundCalculateSizes);
        else window.removeEventListener("resize", this.#boundCalculateSizes);
    }

    // Public API

    /**
     * Start animation
     * @returns {AnimateSprite} - plugin instance
     */
    play(){
        if ( !this.#animation.isAnimating ) this.#animation.play();
        return this;
    }
    /**
     * Stop animation
     * @returns {AnimateSprite} - plugin instance
     */
    stop(){
        this.#animation.stop();
        return this;
    }
    /**
     * Toggle between start and stop
     * @returns {AnimateSprite} - plugin instance
     */
    toggle(){
        if (!this.#animation.isAnimating) this.play();
            else this.stop();
        return this;
    }
    /**
     * Show the next frame
     * @returns {AnimateSprite} - plugin instance
     */
    next(){
        this.stop();
        this.#changeFrame( this.#animation.getNextFrame(1) );
        return this;
    }
    /**
     * Show the previous frame
     * @returns {AnimateSprite} - plugin instance
     */
    prev(){
        this.stop();
        this.#changeFrame( this.#animation.getNextFrame(1, !this.#settings.reverse) );
        return this;
    }
    /**
     * Set frame (without animation)
     * @param {number} frameNumber - Number of the frame to show
     * @returns {AnimateSprite} - plugin instance
     */
    setFrame(frameNumber){
        this.stop();
        this.#changeFrame( normalizeFrameNumber(frameNumber, this.#settings.frames) );
        return this;
    }
    /**
     * Starts the animation, that plays until the specified frame number
     * @param {number} frameNumber - Target frame number
     * @param {Object} [options] - Options
     * @param {boolean} [options.shortestPath=false] - If set to true and loop enabled, will use the shortest path
     * @returns {AnimateSprite} - plugin instance
     */
    playTo(frameNumber, options){
        frameNumber = normalizeFrameNumber(frameNumber, this.#settings.frames);

        const innerPathDistance = Math.abs(frameNumber - this.#data.currentFrame), // not crossing edge frames
            outerPathDistance = this.#settings.frames - innerPathDistance, // crossing edges frames
            shouldUseOuterPath = this.#settings.loop && options?.shortestPath && (outerPathDistance < innerPathDistance);

        if ( !shouldUseOuterPath ) { // Inner path (default)
            // long conditions to make them more readable
            if (frameNumber > this.#data.currentFrame) this.setReverse(false); // move forward
            else this.setReverse(true); // move backward
        } else { // Outer path
            if (frameNumber < this.#data.currentFrame) this.setReverse(false); // move forward
            else this.setReverse(true); // move backward
        }

        return this.playFrames( (shouldUseOuterPath) ? outerPathDistance : innerPathDistance );
    }
    /**
     * Starts the animation in the current direction with the specified number of frames in queue
     * @param {number} [numberOfFrames=0] - Number of frames to play
     * @returns {AnimateSprite} - plugin instance
     */
    playFrames(numberOfFrames = 0){
        numberOfFrames = Math.floor(numberOfFrames);
        if (numberOfFrames < 0) return this.stop(); //empty animation

        this.#animation.framesLeftToPlay = numberOfFrames;
        return this.play();
    }
    /**
     * Change the direction of the animation. Alias to <b>setOption('reverse', true)</b>
     * @param {boolean} [reverse=true] True for backward animation, false for forward
     * @returns {AnimateSprite} - plugin instance
     */
    setReverse(reverse = true){
        this.#settings.reverse = !!reverse;
        return this;
    }
    /** Get current reverse option. Alias to <b>getOption('reverse')</b>
     * @returns {boolean} - reverse or not
     */
    getReverse() { return this.#settings.reverse; }
    /**
     * Calculate new dimensions (sprite element and frame), this function should be called if element size was changed by a script.
     * Called automatically after page resize
     * @returns {AnimateSprite} - plugin instance
     */
    updateSizes(){
        this.#calculateSizes();
        return this;
    }
    /**
     * Returns option value
     * @param {string} option - Option name. All options are allowed.
     * @returns {*} - Option value
     */
    getOption(option){
        const allowedOptions = getSettingsKeys();
        if (allowedOptions.includes(option)) {
            return this.#settings[option];
        } else {
            console.warn(`${option} is not allowed in getOption`);
        }
    }
    /**
     * Set new option value
     * @param {string} option - Option name. All options are allowed.
     * @param {*} value - new value
     * @returns {AnimateSprite} - plugin instance
     */
    setOption(option, value) {
        const allowedOptions = getSettingsKeys();
        if (allowedOptions.includes(option)) {
            this.#settings[option] = value;
            if ( option === "width" || option === "height" || option === "frames" || option === "cols" ) {
                this.#settings[option] = parseInt(value, 10); // additional validation
                this.#calculateSizes();
            }
            else if ( option === "frameTime" || option === "duration" || option === "fps" ) {
                this.#settings.frameTime = this.#settings.duration = this.#settings.fps = false; // Reset
                this.#settings[option] = parseInt(value, 10); // set again because of reset
                this.#animation.updateDuration();
            } else if ( option === "draggable" ) this.#toggleDrag(value);
            else if (option === 'dragModifier') this.#settings.dragModifier = Math.abs(+value);
        } else {
            console.warn(`${option} is not allowed in setOption`);
        }
        return this;
    }
    /** @returns {number} - current frame number */
    getCurrentFrame() { return this.#data.currentFrame; }
    /** @returns {boolean} - animating or not */
    isAnimating() { return this.#animation.isAnimating; }
    /**
     * Stop the animation and return to the first frame
     * @returns @returns {AnimateSprite} - plugin instance
     */
    reset(){
        this.stop();
        this.#changeFrame( normalizeFrameNumber(1, this.#settings.frames) );
        return this;
    }
    /**
     * Stop animation, remove event listeners. Method doesn't remove sprite element from the DOM
     */
    destroy() {
        this.stop();
        this.#toggleDrag(false);
        this.#toggleResizeHandler(false);
    }

}

// todo can't import typedef from another file because it won't add type to d.ts, check this in the future

/**
 * @typedef {object} PluginOptions
 * @property {number} width - Width of one frame in sprite
 * @property {number} height - Height of one frame in sprite
 * @property {number} frames - Number of frames
 * @property {number|false} [cols=false] - Number of cols if more than 1 row
 * @property {boolean} [loop=false] - Whether to loop the animation
 * @property {boolean} [autoplay=false] - Autoplay
 * @property {number|false} [frameTime] - ms, time between frames
 * @property {number|false} [duration] - ms, total time, alternative to frameTime
 * @property {number|false} [fps=24] - fps, alternative to frameTime
 * @property {boolean} [reverse=false] - Reverse direction of animation
 * @property {number|boolean} [draggable=false] - Draggable by mouse or touch
 * @property {boolean} [inversion=false] - Inversion changes drag direction
 * @property {number} [dragModifier=1] - Sensitivity factor for user interaction. Only positive numbers are allowed
 * @property {'pageScrollTimer' | 'preventPageScroll' | 'allowPageScroll'} [touchScrollMode="pageScrollTimer"]
 * - Page scroll behavior with touch events (preventPageScroll,allowPageScroll, pageScrollTimer)
 * @property {number} [pageScrollTimerDelay=1500] - Time in ms when touch scroll will be disabled during interaction
 * if options.touchScrollMode = "pageScrollTimer"
 * @property {function(AnimateSprite):void} [onAnimationEnd] - Occurs when animation has ended
 * @property {function(AnimateSprite):void} [onAfterFrameChange] - Occurs after the frame has switched
 */

